src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,304 @@
+
+    for (unsigned int i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].aId == id) {
+        return true;
+      }
+    }
+    
+    return false;
+  }
+
+  // Get neuron by id.
+  NeuronGene GetNeuronById(unsigned int id) {
+    assert(HasNeuronId(id));
+
+    for (unsigned int i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].aId == id) {
+        return aNeuronGenes[i];
+      }
+    }
+  }
+
+  // Get neuron index by id.
+  unsigned int GetNeuronIndex(unsigned int id) const {
+    for(unsigned int i=0; i < NumNeuron(); ++i) {
+        if (aNeuronGenes[i].Id() == id) {
+            return i;
+        }
+    }
+
+    return -1;
+  }
+
+  // Calculate Neuron depth.
+  unsigned int NeuronDepth(unsigned int id) {
+    // TODO: if contains loop in network.
+
+    // Find all links that output to this neuron id.
+    std::vector<int> inputLinksIndex;
+    for (unsigned int i=0; i<NumLink(); ++i) {
+      if (aLinkGenes[i].ToNeuronId() == id) {
+        inputLinksIndex.push_back(i);
+      }
+    }
+
+    // INPUT or BIAS or isolated neurons.
+    if (inputLinksIndex.size() == 0) {
+      return 0;
+    }
+
+    // Recursively get neuron depth.
+    std::vector<unsigned int> depths(inputLinksIndex.size());
+    for (unsigned int i=0; i<inputLinksIndex.size(); ++i) {
+      depths[i] = NeuronDepth(aLinkGenes[inputLinksIndex[i]].FromNeuronId());
+    }
+    unsigned int maxInputDepth = *(std::max_element(std::begin(depths),
+                                 std::end(depths)));
+    return (maxInputDepth + 1);
+  }
+
+  // Calculate Genome depth.
+  // It is the max depth of all output neuron genes.
+  unsigned int GenomeDepth() {
+    unsigned int numNeuron = NumNeuron();
+
+    // If empty genome.
+    if (numNeuron == 0) {
+      aDepth = 0;
+      return aDepth;
+    }
+
+    // If no hidden neuron, depth is 1.
+    if (aNumInput + aNumOutput == numNeuron) {
+      aDepth = 1;
+      return aDepth;
+    }
+    
+    // Find all OUTPUT neuron id.
+    std::vector<unsigned int> outputNeuronsId;
+    for (unsigned int i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].Type() == OUTPUT) {
+        outputNeuronsId.push_back(aNeuronGenes[i].Id());
+      }
+    }
+
+    // Get max depth of all output neurons.
+    unsigned int genomeDepth = 0;
+    for (unsigned int i=0; i<outputNeuronsId.size(); ++i) {
+      unsigned int outputNeuronDepth = NeuronDepth(outputNeuronsId[i]);
+      if (outputNeuronDepth > genomeDepth) {
+        genomeDepth = outputNeuronDepth;
+      }
+    }
+    aDepth = genomeDepth;
+
+    return aDepth;
+  }
+
+  // Set neurons' input and output to zero.
+  void Flush() {
+    for (unsigned int i=0; i<aNeuronGenes.size(); ++i) {
+      aNeuronGenes[i].aActivation = 0;
+      aNeuronGenes[i].aInput = 0;
+    }
+  }
+
+  // Activate genome. The last dimension of input is always 1 for bias. 0 means no bias.
+  void Activate(std::vector<double>& input) {
+    assert(input.size() == aNumInput);
+    Flush();
+
```

Comment:
If we reset the activation at the beginning, we can't use the activation at time  (t - 1) from a unit that has a recurrent connection, right? I'm not sure we have to reset the activation at all, we could just overwrite the existing.